planner/core/memtable_predicate_extractor.go

```diff
@@ -0,0 +1,204 @@
+			return "", nil
+		}
+		results = append(results, constant.Value)
+	}
+	return name.ColName.L, results
+}
+
+func (helper extractHelper) extractColEqConsExpr(extractCols map[int64]*types.FieldName, expr *expression.ScalarFunction) (string, types.Datum) {
+	args := expr.GetArgs()
+	col, isCol := args[0].(*expression.Column)
+	if !isCol {
+		return "", types.Datum{}
+	}
+	name, found := extractCols[col.UniqueID]
+	if !found {
+		return "", types.Datum{}
+	}
+	// The `rhs` of EQ expression must be a constant
+	// SELECT * FROM t1 WHERE c = 'rhs'
+	constant, ok := args[1].(*expression.Constant)
+	if !ok || constant.DeferredExpr != nil || constant.ParamMarker != nil {
+		return "", types.Datum{}
+	}
+	return name.ColName.L, constant.Value
+}
+
+// ClusterConfigTableExtractor is used to extract some predicates of `TIDB_CLUSTER_CONFIG`
+type ClusterConfigTableExtractor struct {
+	extractHelper
+
+	// SkipRequest means the where cluase always false, we don't need to request any component
+	SkipRequest bool
+
+	// NodeTypes represents all components types we should send request to.
+	// e.g:
+	// 1. SELECT * FROM tidb_cluster_config WHERE type='tikv'
+	// 2. SELECT * FROM tidb_cluster_config WHERE type in ('tikv', 'tidb')
+	NodeTypes set.StringSet
+
+	// Addresses represents all components addresses we should send request to.
+	// e.g:
+	// 1. SELECT * FROM tidb_cluster_config WHERE address='192.168.1.7:2379'
+	// 2. SELECT * FROM tidb_cluster_config WHERE type in ('192.168.1.7:2379', '192.168.1.9:2379')
+	Addresses set.StringSet
+}
+
+// Extract implements the MemTablePredicateExtractor Extract interface
+func (e *ClusterConfigTableExtractor) Extract(schema *expression.Schema, names []*types.FieldName, predicates []expression.Expression) []expression.Expression {
+	remained := make([]expression.Expression, 0, len(predicates))
+	// All columns can be pushed down to the memory table `TIDB_CLUSTER_CONFIG`
+	const (
+		ColNameType    = "type"
+		ColNameAddress = "address"
+	)
+	extractCols := make(map[int64]*types.FieldName)
+	for i, name := range names {
+		if ln := name.ColName.L; ln == ColNameType || ln == ColNameAddress {
+			extractCols[schema.Columns[i].UniqueID] = name
+		}
+	}
+	// We don't support change the schema of memory table, so something must
+	// be wrong if the `type/address` columns cannot be found in the schema.
+	// The purpose of the following assert is used to make test failure because
+	// it will never happen if our release passing all test cases.
+	if len(extractCols) != 2 {
+		panic(fmt.Sprintf("push down columns `type/address` not found in schema, got: %+v", extractCols))
+	}
+
+	e.NodeTypes = set.NewStringSet()
+	e.Addresses = set.NewStringSet()
+
+	// We should use INTERSECTION of sets because of the predicates is CNF array
+	for _, expr := range predicates {
+		var extracted bool
+		switch x := expr.(type) {
+		case *expression.ScalarFunction:
+			switch x.FuncName.L {
+			case ast.EQ:
```

Comment:
We can combine the two branches like this:
```go
var extractFunc func(extractCols map[int64]*types.FieldName, expr *expression.ScalarFunction) (string, []types.Datum)
switch x.FuncName.L {
case ast.EQ:
	extractFunc = e.extractColEqConsExpr
case ast.In
	extractFunc = e.extractColInConsExpr
}
colName, datums := extractFunc(extractCols, x)
// then deal the datums and colName.
```

Reply:
done